### PR TITLE
PHPLIB-1596 Fully describe `$search`, `$searchMeta` and `$vectorSearch` stages

### PR DESCRIPTION
**Pull request by GromNaN**
Fix PHPLIB-1596

The parameters of the `$search` and `$searchMeta` stages are currently not described. Changing the generic property object into a full list of named arguments would be a breaking change after the first release of this code, so it's now or never.

For the tests, you can see the deep structure of options is not described. That's something we an do later.

- [x] https://github.com/mongodb/mongo-php-library/pull/1540
- [x] https://github.com/mongodb/mongo-php-library/pull/1541

### CODE REVIEW

File: generator/composer.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
     "repositories": [
         {
             "type": "path",
-            "url": "../",
+            "url": "..",
```
Comment (jmikola) [inline generator/composer.json:7]: Was the trailing slash just redundant?

File: generator/src/OperatorTestGenerator.php [hunk 2]
```diff
@@ -145,6 +147,8 @@ private function convertYamlTaggedValues(mixed $object): mixed
                 'bson_decimal128' => new Decimal128($value),
                 'bson_utcdatetime' => new UTCDateTime(is_numeric($value) ? $value : new DateTimeImmutable($value)),
                 'bson_binary' => new Binary(base64_decode($value)),
+                'bson_objectId' => new ObjectId($value),
+                'bson_uuid' => new Binary(pack('h*', str_replace('-', '', $value)), Binary::TYPE_UUID),
```
Comment (jmikola) [inline generator/src/OperatorTestGenerator.php]: Would [`hex2bin()`](https://www.php.net/manual/en/function.hex2bin.php) be more appropriate here? It's not immediately clear to me if the `h` format (low nibble first) for [`pack()`](https://www.php.net/pack) is correct.

I'll note that [`bin2hex()`](https://www.php.net/bin2hex) produces a string with the high nibble first.

File: generator/src/OperatorTestGenerator.php [hunk 3]
```diff
@@ -158,6 +162,7 @@ private function convertYamlTaggedValues(mixed $object): mixed
         }
 
         if (is_object($object)) {
+            $object = clone $object;
```
Comment (jmikola) [inline generator/src/OperatorTestGenerator.php:165]: Was this a bug fix? It looks unrelated to implementing search operators.

File: tests/Builder/Stage/VectorSearchStageTest.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+use MongoDB\Builder\Pipeline;
+use MongoDB\Builder\Query;
+use MongoDB\Builder\Stage;
+use MongoDB\Tests\Builder\PipelineTestCase;
+
+/**
+ * Test $vectorSearch stage
+ */
+class VectorSearchStageTest extends PipelineTestCase
```
Comment (jmikola) [inline tests/Builder/Stage/VectorSearchStageTest.php:15]: Are these files generated, or is that only `tests/Builder/Stage/Pipelines.php` (which has a header comment indicating as much). None of these appear to be automatically hidden in the code review, so I wonder if the repo config needs an update for changed paths.

File: tests/Builder/Search/Pipelines.php [hunk 5]
```diff
@@ -0,0 +1,2814 @@
+<?php
+
+/**
+ * THIS FILE IS AUTO-GENERATED. ANY CHANGES WILL BE LOST!
```
Comment (jmikola) [inline tests/Builder/Search/Pipelines.php:4]: Noted this file is auto-generated but GitHub only suppressed it because of its size. The path may need to be added in the repo config.

File: tests/Builder/Search/InOperatorTest.php [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+use DateTimeImmutable;
+use MongoDB\BSON\ObjectId;
+use MongoDB\BSON\UTCDateTime;
+use MongoDB\Builder\Pipeline;
+use MongoDB\Builder\Search;
+use MongoDB\Builder\Stage;
+use MongoDB\Tests\Builder\PipelineTestCase;
+
+/**
+ * Test in search
```
Comment (jmikola) [inline tests/Builder/Search/InOperatorTest.php:16]: If these descriptions are hand-written, it may be preferable to use the `$` prefix so they actually read as the operator names. This could be misread as a sentence, leaving the reader to ask, "test _what_ in the search?"

Consider:

> Test $in search operator

File: tests/Builder/Stage/VectorSearchStageTest.php [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+namespace MongoDB\Tests\Builder\Stage;
+
+use MongoDB\Builder\Pipeline;
+use MongoDB\Builder\Query;
+use MongoDB\Builder\Stage;
+use MongoDB\Tests\Builder\PipelineTestCase;
+
+/**
+ * Test $vectorSearch stage
```
Comment (jmikola) [inline tests/Builder/Stage/VectorSearchStageTest.php:13]: I appreciate using `$` to prefix the stage names here.

File: generator/config/expressions.php [hunk 8]
```diff
@@ -168,4 +172,12 @@
     'GeoPoint' => [
         'acceptedTypes' => [...$bsonTypes['object']],
     ],
+
+    // Search
+    'searchPath' => [
```
Comment (jmikola) [inline generator/config/expressions.php:177]: Why do these keys start with lowercase letters? It looks inconsistent with the keys above.

Also, the comment on the line above may not be necessary since both keys start with "search". If the comment is intentional to distinguish these two array elements from those above, that's fine. I expect answering my first question above will clear up any misunderstanding.

File: generator/config/search/autocomplete.yaml [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+            - searchPath
+    -
+        name: query
+        type:
+            - string
+    -
+        name: tokenOrder
+        optional: true
+        type:
+            - string #tokenOrder
```
Comment (jmikola) [inline generator/config/search/autocomplete.yaml]: Does this comment add any value? I assume it may be referring to:

```
"tokenOrder": "any|sequential",
```

I assume you don't want to define an enum for that, nor enforce it client-side, but are the possible values documented anywhere in the builder?

File: generator/config/search/embeddedDocument.yaml [hunk 10]
```diff
@@ -0,0 +1,155 @@
+# $schema: ../schema.json
+name: embeddedDocument
+link: 'https://www.mongodb.com/docs/atlas/atlas-search/embeddedDocument/'
```
Comment (jmikola) [inline generator/config/search/embeddedDocument.yaml]: This looks like a 404. Also, I need to reiterate how annoying it is that our docs website redirects 404s to https://www.mongodb.com/docs/404/ instead of leaving the original URL as-is. I've reported this in the past and the feedback wasn't acted upon :disappointed: 

I think this should be https://www.mongodb.com/docs/atlas/atlas-search/embedded-document/

File: generator/config/search/compound.yaml [hunk 11]
```diff
@@ -0,0 +1,156 @@
+# $schema: ../schema.json
+name: compound
+link: 'https://www.mongodb.com/docs/atlas/atlas-search/equals/'
```
Comment (jmikola) [inline generator/config/search/compound.yaml]:
```suggestion
link: 'https://www.mongodb.com/docs/atlas/atlas-search/compound/'
```

File: generator/config/search/equals.yaml [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+        type:
+            - searchPath
+    -
+        name: value
+        type:
+            - binData
+            - bool
+            - date
+            - objectId
+            - 'null'
```
Comment (jmikola) [inline generator/config/search/equals.yaml:21]: Is this necessary to overcome some YAML behavior, where `null` would be interpreted as `~`?

File: generator/config/search/exists.yaml [hunk 13]
```diff
@@ -0,0 +1,56 @@
+# $schema: ../schema.json
+name: exists
+link: 'https://www.mongodb.com/docs/atlas/atlas-search/equals/'
```
Comment (jmikola) [inline generator/config/search/exists.yaml]:
```suggestion
link: 'https://www.mongodb.com/docs/atlas/atlas-search/exists/'
```

File: generator/config/search/exists.yaml [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+type:
+    - searchOperator
+encode: object
+description: |
+    The equals operator checks whether a field matches a value you specify.
```
Comment (jmikola) [inline generator/config/search/exists.yaml]: This looks like copypasta.

File: generator/config/search/facet.yaml [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    - searchOperator # should be searchCollector
+encode: object
+description: |
+    The facet collector groups results by values or ranges in the specified
+    faceted fields and returns the count for each of those groups.
+arguments:
+    -
+        name: facets
+        type:
+            - object # of facetDefinition
```
Comment (jmikola) [inline generator/config/search/facet.yaml]: You previously used "of <type>" for an `array`. Was this intentional?

I assume it may be because it's a hash of facet names to definitions. Maybe change this to "# map of facetDefinition"?

File: generator/config/search/geoShape.yaml [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+    geometry if indexShapes is set to true in the index definition.
+arguments:
+    -
+        name: path
+        type:
+            - searchPath
+    -
+        name: relation
+        type:
+            - string # contains | disjoint | intersects | within
```
Comment (jmikola) [inline generator/config/search/geoShape.yaml:18]: Curious why you enumerated the possible values here, but not for the tokenOrder string earlier.

File: generator/config/search/geoShape.yaml [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+                $search:
+                    geoShape:
+                        relation: 'disjoint'
+                        geometry:
+                            type: 'Polygon'
+                            coordinates:
+                                -
+                                    -
+                                        - -161.323242
+                                        - 22.512557
```
Comment (jmikola) [inline generator/config/search/geoShape.yaml]: Would it be more readable to inline these, like `[ -161.323242, 22.512557 ]`? That might reduce a lot of lines in this file.

If there's a concern that it's generated and might conflict with future syncs, no worries leaving as-is. I'm not familiar with how these are imported.

File: generator/config/search/geoWithin.yaml [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+        pipeline:
+            -
+                $search:
+                    geoWithin:
+                        geometry:
+                            type: 'Polygon'
+                            coordinates:
+                                -
+                                    -
+                                        - -161.323242
```
Comment (jmikola) [inline generator/config/search/geoWithin.yaml]: Same inlining suggestion for this file.

**Review by jmikola — COMMENTED** [review]
Taking a first pass and will leave more feedback on other YAML files later.

File: generator/config/search/autocomplete.yaml (hunk 9 above)
Comment (GromNaN) [reply generator/config/search/autocomplete.yaml]: Updating the comment, as I don't have implemented this enum for now.
```suggestion
            - string # any|sequential
```

File: generator/config/expressions.php (hunk 8 above)
Comment (GromNaN) [reply generator/config/expressions.php:177]: I think that when I did this initially, there was a distinction between “scalar” and “object” types in my mind. But that doesn't make sense. I need to harmonize this. It has no impact on the generated code.

File: generator/config/search/embeddedDocument.yaml (hunk 10 above)
Comment (GromNaN) [reply generator/config/search/embeddedDocument.yaml]:
```suggestion
link: 'https://www.mongodb.com/docs/atlas/atlas-search/embedded-document/'
```

File: generator/config/search/equals.yaml (hunk 12 above)
Comment (GromNaN) [reply generator/config/search/equals.yaml:21]: Exactly, the unquoted `null` string is the NULL value in Yaml. But it seems that I can disable this behavior: https://github.com/symfony/yaml/blob/3ced3f29e4f0d6bce2170ff26719f1fe9aacc671/CHANGELOG.md#440

Let's stay safe and keep the quotes, if other scripts use this Yaml files.

File: generator/config/search/exists.yaml (hunk 14 above)
Comment (GromNaN) [reply generator/config/search/exists.yaml]:
```suggestion
    The exists operator tests if a path to a specified indexed field name exists in a document.
```

File: generator/config/search/facet.yaml (hunk 15 above)
Comment (GromNaN) [reply generator/config/search/facet.yaml]: Updated, thanks.
This comments are a kind of todo list to improve typing.
```suggestion
            - object # map of facetDefinition
```

File: generator/config/search/geoShape.yaml (hunk 16 above)
Comment (GromNaN) [reply generator/config/search/geoShape.yaml:18]: Updating tokenOrder

File: generator/config/search/geoShape.yaml (hunk 17 above)
Comment (GromNaN) [reply generator/config/search/geoShape.yaml]: It's generated by my `js2yaml.html` file. I've made a few editions in some places, so why not...

File: generator/src/OperatorTestGenerator.php (hunk 2 above)
Comment (GromNaN) [reply generator/src/OperatorTestGenerator.php]: Thanks.
```suggestion
                'bson_uuid' => new Binary(hex2bin(str_replace('-', '', $value)), Binary::TYPE_UUID),
```

File: generator/src/OperatorTestGenerator.php (hunk 3 above)
Comment (GromNaN) [reply generator/src/OperatorTestGenerator.php:165]: This is kind of but that doesn't affect the current code. When calling `$pipeline = $this->convertYamlTaggedValues($test->pipeline);`, the original `$test->pipeline` is modified, which is not expected.

I hit this issue because I hacked into this file to generate the pipeline of test methods with some dumb implementation. I did not commit that because that was not good.

File: tests/Builder/Search/Pipelines.php (hunk 5 above)
Comment (GromNaN) [reply tests/Builder/Search/Pipelines.php:4]: I'll fix the .gitattribute config.

File: tests/Builder/Stage/VectorSearchStageTest.php (hunk 4 above)
Comment (GromNaN) [reply tests/Builder/Stage/VectorSearchStageTest.php:15]: The test files are generated with almost empty method body, then I manually write every aggregation pipeline using the builder syntax. That why this files are not flagged as "generated".

File: generator/composer.json (hunk 1 above)
Comment (GromNaN) [reply generator/composer.json:7]: That was not necessary.

**Review by GromNaN — COMMENTED** [review]
Thanks for this first review. I'll commit the fixes.

File: generator/src/OperatorTestGenerator.php (hunk 3 above)
Comment (jmikola) [reply generator/src/OperatorTestGenerator.php:165]: Oh, I originally missed that this was in the generator sources. Glad it's fixed and I see no need for a PHPLIB ticket to document this since the code is internal.

File: generator/src/OperatorTestGenerator.php [hunk 19]
```diff
@@ -145,6 +147,8 @@ private function convertYamlTaggedValues(mixed $object): mixed
                 'bson_decimal128' => new Decimal128($value),
                 'bson_utcdatetime' => new UTCDateTime(is_numeric($value) ? $value : new DateTimeImmutable($value)),
                 'bson_binary' => new Binary(base64_decode($value)),
+                'bson_objectId' => new ObjectId($value),
+                'bson_uuid' => new Binary(hex2bin(str_replace('-', '', $value)), Binary::TYPE_UUID),
```
Comment (jmikola) [inline generator/src/OperatorTestGenerator.php:151]: Because I was curious, I went back and compared the original code you had with `pack('h*', ...)`. See below:

```php
<?php

// Initialized project with: composer require ramsey/uuid
require 'vendor/autoload.php';

use Ramsey\Uuid\Uuid;
use MongoDB\BSON\Binary;

$uuid = Uuid::uuid4();
var_dump($uuid);

$bytes = $uuid->getBytes();
$str = $uuid->toString();

$bin = new Binary($bytes, Binary::TYPE_UUID);
var_dump(bin2hex($bin));

$bin2 = new Binary(hex2bin(str_replace('-','',$str)), Binary::TYPE_UUID);
var_dump($bin == $bin2);

$bin3 = new Binary(pack('h*', str_replace('-','',$str)), Binary::TYPE_UUID);
var_dump($bin == $bin3);

$bin4 = new Binary(pack('H*', str_replace('-','',$str)), Binary::TYPE_UUID);
var_dump($bin == $bin4);
```

Output:

```
$ php foo.php 
class Ramsey\Uuid\Lazy\LazyUuidFromString#4 (2) {
  private ?Ramsey\Uuid\UuidInterface $unwrapped =>
  NULL
  private string $uuid =>
  string(36) "3a6224c2-f648-4978-b510-610c04fbbcd6"
}
string(32) "3a6224c2f6484978b510610c04fbbcd6"
bool(true)
bool(false)
bool(true)
```

So it looks like `h` (hex string, low nibble first) wouldn't have worked, but `H` (high nibble first) would have. `hex2bin()` is totally fine, though, and easier to read.

File: src/Builder/Encoder/OperatorEncoder.php [hunk 20]
```diff
@@ -129,6 +132,34 @@ private function encodeAsObject(OperatorInterface $value): stdClass
             : $this->wrap($value, $result);
     }
 
+    private function encodeAsSearch(OperatorInterface $value): stdClass
+    {
+        assert($value instanceof SearchStage || $value instanceof SearchMetaStage);
```
Comment (jmikola) [inline src/Builder/Encoder/OperatorEncoder.php]: Is there a reason these methods use `assert()` instead of strict type hints?

File: src/Builder/Encoder/OperatorEncoder.php [hunk 21]
```diff
@@ -129,6 +132,34 @@ private function encodeAsObject(OperatorInterface $value): stdClass
...
+            // Skip optional arguments. If they have a default value, it is resolved by the server.
+            if ($val === Optional::Undefined) {
+                continue;
+            }
+
+            // Merge the operator properties into the result object
+            if ($key === 'operator') {
+                $val = $this->recursiveEncode($val);
+                assert($val instanceof stdClass);
+                foreach (get_object_vars($val) as $subkey => $subval) {
```
Comment (jmikola) [inline src/Builder/Encoder/OperatorEncoder.php]: If `$val` is a stdClass, do you even need `get_object_vars()`? [Basic object iteration](https://www.php.net/manual/en/language.oop5.iterations.php) rules should apply. No objections if you feel this is more explicit/consistent.

Separate question: is the logic here also frequently applied in other operators? Or is it unique in its unwrapping and merging of the `operator` object into the parent context?

Is there any concern that operator keys (i.e. `$subkey` values) might conflict with a top-level `$key`?

File: generator/config/stage/search.yaml [hunk 22]
```diff
@@ -3,16 +3,92 @@ name: $search
...
+        description: |
+            Operator to search with.  You can provide a specific operator or use
+            the compound operator to run a compound query with multiple operators.
+    -
+        name: index
+        optional: true
+        type:
+            - string
+        description: |
+            Name of the Atlas Search index to use. If omitted, defaults to default.
```
Comment (jmikola) [inline generator/config/stage/search.yaml]:
```suggestion
            Name of the Atlas Search index to use. If omitted, defaults to "default".
```

Would be helpful to clarify the string literal here.

File: generator/config/search/queryString.yaml [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+        type:
+            - searchPath
+    -
+        name: query
+        type:
+            - string
+
+# The various example from the doc are variations of the "query" parameter
+# this is not pertinent for testing the aggregation builder, unless we create
+# a queryString builder.
```
Comment (jmikola) [inline generator/config/search/queryString.yaml:21]: Let's not :wink:

File: generator/config/expressions.php (hunk 8 above)
Comment (GromNaN) [reply generator/config/expressions.php:177]: [PHPLIB-1598](https://jira.mongodb.org/browse/PHPLIB-1598)

File: tests/Builder/Search/InOperatorTest.php (hunk 6 above)
Comment (GromNaN) [reply tests/Builder/Search/InOperatorTest.php:16]: In fact, the search operators are not prefixed by the dollar sign 🤦.
I'll do a following PR to add quotes as this will impact all test classes (or maybe remove the comment).

File: src/Builder/Encoder/OperatorEncoder.php (hunk 20 above)
Comment (GromNaN) [reply src/Builder/Encoder/OperatorEncoder.php]: Indeed, fixed.

File: src/Builder/Encoder/OperatorEncoder.php (hunk 21 above)
Comment (GromNaN) [reply src/Builder/Encoder/OperatorEncoder.php]: > If `$val` is a stdClass, do you even need `get_object_vars()`? [Basic object iteration](https://www.php.net/manual/en/language.oop5.iterations.php) rules should apply. No objections if you feel this is more explicit/consistent.

For the time being, in removed the `get_object_vars`.

> Separate question: is the logic here also frequently applied in other operators? Or is it unique in its unwrapping and merging of the `operator` object into the parent context?
> Is there any concern that operator keys (i.e. `$subkey` values) might conflict with a top-level `$key`?

I opened a PR to solve the need of merging the subproperties into the parent object: https://github.com/mongodb/mongo-php-library/pull/1540

**Review by GromNaN — COMMENTED** [review]
Let's review https://github.com/mongodb/mongo-php-library/pull/1540 before going further.

File: src/Builder/Encoder/OperatorEncoder.php (hunk 20 above)
Comment (GromNaN) [reply src/Builder/Encoder/OperatorEncoder.php]: Method removed.

File: tests/Builder/Search/Pipelines.php (hunk 5 above)
Comment (GromNaN) [reply tests/Builder/Search/Pipelines.php:4]: There is this line in the `.gitattribute` file that should have detected this file.
```
/tests/Builder/*/Pipelines.php linguist-generated=true
```

File: generator/config/stage/search.yaml [hunk 24]
```diff
@@ -3,16 +3,93 @@ name: $search
...
+            - object
+        description: |
+            Document that specifies the fields to sort the Atlas Search results by in ascending or descending order.
+    -
+        name: returnStoredSource
+        optional: true
+        type:
+            - bool
+        description: |
+            Flag that specifies whether to return the stored source document in the search results. If omitted, defaults to false.Flag that specifies whether to perform a full document lookup on the backend database or return only stored source fields directly from Atlas Search.
```
Comment (jmikola) [inline generator/config/stage/search.yaml]: Missing spacing between the sentences here. Perhaps a copy/paste error?

File: generator/config/stage/search.yaml [hunk 25]
```diff
@@ -3,16 +3,93 @@ name: $search
...
+            Specifies the highlighting options for displaying search terms in their original context.
+    -
+        name: concurrent
+        optional: true
+        type:
+            - bool
+        description: |
+            Parallelize search across segments on dedicated search nodes.
+            If you don't have separate search nodes on your cluster, 
+            Atlas Search ignores this flag. If omitted, defaults to false.
```
Comment (jmikola) [inline generator/config/stage/search.yaml:43]: Elsewhere, I recall you specifying `default: <value>` in the YAML configs. Should that be done here? Likewise for the `index` option above and some other params below with default values.

File: generator/config/stage/searchMeta.yaml [hunk 26]
```diff
@@ -3,16 +3,34 @@ name: $searchMeta
...
+        description: |
+            Operator to search with.  You can provide a specific operator or use
+            the compound operator to run a compound query with multiple operators.
+    -
+        name: index
+        optional: true
+        type:
+            - string
+        description: |
+            Name of the Atlas Search index to use. If omitted, defaults to default.
```
Comment (jmikola) [inline generator/config/stage/searchMeta.yaml:25]: Should `default:` be used here?

File: generator/config/stage/vectorSearch.yaml [hunk 27]
```diff
@@ -0,0 +1,129 @@
...
+        description: |
+            This is required if numCandidates is omitted.
+            - false to run ANN search
+            - true to run ENN search
+    -
+        name: filter
+        optional: true
+        type:
+            - query
+        description: |
```
Comment (jmikola) [inline generator/config/stage/vectorSearch.yaml:46]: Should these descriptions be empty? Likewise for `numCandidates` below.

File: generator/config/stage/search.yaml (hunk 25 above)
Comment (GromNaN) [reply generator/config/stage/search.yaml:43]: The `default` option is only when we want a value to be set by default. This is only when a value is required, otherwise we rely on the default value server-side.

File: generator/config/stage/searchMeta.yaml (hunk 26 above)
Comment (GromNaN) [reply generator/config/stage/searchMeta.yaml:25]: See https://github.com/mongodb/mongo-php-library/pull/1537#discussion_r1868405362

**Review by jmikola — APPROVED** [review]
